django_ses/signals.py

```diff
@@ -1,9 +1,31 @@
 from django.dispatch import Signal
 
+from django_ses import models, settings
+
 # The following fields are used from the 3 signals below: mail_obj, bounce_obj, raw_message
 bounce_received = Signal()
 complaint_received = Signal()
 delivery_received = Signal()
 send_received = Signal()
 open_received = Signal()
 click_received = Signal()
+
+
+def bounce_handler(sender, mail_obj, bounce_obj, raw_message, *args, **kwargs):
+    if not settings.AWS_SES_ENABLE_BOUNCE_BLACKLIST:
+        return
+
+    from django_ses.utils import get_permanent_bounced_emails_from_bounce_obj
+    bounced_recipients = get_permanent_bounced_emails_from_bounce_obj(bounce_obj)
+    for email in bounced_recipients:
+        models.BlacklistedEmail.objects.get_or_create(email=email.lower())
```

Comment:
Would prefer to bulk-create these, can we do that?

https://docs.djangoproject.com/en/5.1/ref/models/querysets/#bulk-create

Reply:
We could do it like this:

```
models.BlacklistedEmail.objects.bulk_create([
    models.BlacklistedEmail(email=email.lower()) for email in bounced_recipients
], ignore_conflicts=True)
```

but Oracle doesn't support `ignore_conflicts` (https://docs.djangoproject.com/en/5.1/ref/models/querysets/#bulk-create), which means that the extension will crash every time the same emails is bounced and the app is running on an Oracle database.

Should I use the `bulk_create` anyways?

Reply:
I missed the fact that you're using `get_or_create` meaning we need to check conflicts.

You could first filter out existing recipients using another db lookup and then run `bulk_create`. Let's do it with `ignore_conflicts` anyway just to be safe in case of race conditions.

How do you know it will crash on an Oracle database, instead of ignoring the flag? In theory if we check for conflicts, and the flag is ignored, then it should still run successfully except in rare cases of race conditions that would have created the recipient between the time we check and the time we write.

Reply:
I could do something like

```
27
28 already_existing_emails = BlacklistedEmail.objects.filter(...<the emails received from the bounce>)
29 emails_that_must_be_inserted = <the emails received from the bounce> | already_existing_emails
30 BlacklistedEmail.objects.bulk_insert(emails_that_must_be_inserted, ignore_conflicts=True)
31 
```

but that code has 2 problems:

a) another process can race condition "us" between ` 29` and `30`, which will lead us attempt an insert with already existing emails, this is bad because b)
b) as I said in my previous post, [Oracle doesn't support bulk-create with ignore_conflicts] (https://docs.djangoproject.com/en/5.1/ref/models/querysets/#bulk-create), which will lead to a crash.

We could fix `a)` with table locking, but that would require a lot of hacking (because we'd want to support multiple databases, right?)

Inserting mails one-by-one feels dirty, but I can't figure out any better way of doing it.

![image](https://github.com/user-attachments/assets/7159086e-0163-4369-962f-1ee5a88690ea)

Reply:
I acknowledged the race condition. That's why I said to keep the ignore conflicts flag.

But I don't see where it says it will crash on Oracle if you pass the flag. I understand that it will throw an error if there is a conflict. So maybe that is when we catch the error and fall back to insert one by one?

So changes

1. Filter out dupes first. I know there is still a race condition but it reduces the risk of fallback.
2. Use bulk create with the ignore conflicts flag
3. Fall back to current way if bulk fails.

Reply:
Ooohh, I see now. Sorry, I wasn't understanding you. Okey, let me send a patch for this!

Reply:
Just pushed the change! Can you look at it? I'm not 100% convinced it's good enough, but I'm not sure why.